src/loaders/jobs/taskfarmer/task_generator.py

```diff
@@ -0,0 +1,247 @@
+
+    # Check if the image is already present on the system
+    std_out, std_err, returncode = _run_command(["shifterimg", "images"])
+
+    images = std_out.split("\n")
+    for image in images:
+        parts = image.split()
+        if len(parts) != 6:
+            continue
+        if parts[5] == image_str:
+            print(f"Shifter image {image_name}:{ver} from registry {registry} already exists.")
+            return parts[5]
+
+    # Pull the image from the registry
+    print(f"Fetching Shifter image {image_str}...")
+    std_out, std_err, returncode = _run_command(["shifterimg", "pull", image_str])
+
+    if 'FAILURE' in std_out:
+        raise ValueError(f"Error fetching Shifter image {image_str}.")
+
+    return image_str
+
+
+def _create_shifter_wrapper(job_dir, image_str):
+    """
+    Creates the Shifter wrapper script.
+    """
+
+    # The content of the Shifter wrapper script
+    shifter_wrapper = "#!/bin/bash\n\n"
+    shifter_wrapper += f"image={image_str}\n\n"
+    shifter_wrapper += "if [ $# -lt 1 ]; then\n"
+    shifter_wrapper += '    echo "Error: Missing command argument."\n'
+    shifter_wrapper += '    echo "Usage: shifter_wrapper.sh your-command-arguments"\n'
+    shifter_wrapper += "    exit 1\n"
+    shifter_wrapper += "fi\n\n"
+    shifter_wrapper += "command=\"$@\"\n\n"
+    shifter_wrapper += "echo \"Running shifter --image=$image $command\"\n\n"
+    shifter_wrapper += f"cd $HOME\n"
+    shifter_wrapper += "shifter --image=$image $command\n"  # Run the command in the Shifter environment
+
+    wrapper_file = os.path.join(job_dir, 'shifter_wrapper.sh')
+    with open(wrapper_file, "w") as f:
+        f.write(shifter_wrapper)
+
+    os.chmod(wrapper_file, 0o777)
+
+    return wrapper_file
+
+
+def _create_genome_id_file(genome_ids, genome_id_file):
+    """
+    Create a tab-separated values (TSV) file with a list of genome IDs.
+    """
+    with open(genome_id_file, 'w') as f:
+        f.write("genome_id\n")
+        for genome_id in genome_ids:
+            f.write(f"{genome_id}\n")
+
+
+def _create_task_list(source_data_dir, kbase_collection, load_ver, tool, wrapper_file, job_dir, root_dir):
+    """
+    Create task list file (tasks.txt)
+    """
+    genome_ids = [path for path in os.listdir(source_data_dir) if
+                  os.path.isdir(os.path.join(source_data_dir, path))]
+
+    genome_ids_chunks = [genome_ids[i: i + CHUNK_SIZE] for i in range(0, len(genome_ids), CHUNK_SIZE)]
+
+    task_list = '#!/usr/bin/env bash\n'
+    for idx, genome_ids_chunk in enumerate(genome_ids_chunks):
+
+        task_list += wrapper_file
+
+        genome_id_file = os.path.join(job_dir, f'genome_id_{idx}.tsv')
+        _create_genome_id_file(genome_ids_chunk, genome_id_file)
+
+        if tool == 'checkm2':
+            task_list += f'	--volume={CHECKM2_DB}:/CheckM2_database --volume={root_dir}:/checkm2_root_dir ' \
+                         f'conda run -n checkm2-1.0.0 ' \
```

Comment:
This seems like an issue if the checkm2 or gtdb_tk docker image is updated with a new version of the binary

I'm a little iffy about replacing the command line altogether. Maybe just providing an envvar with the tool to run would be a better approach and let the docker container figure out the rest

Reply:
That would also simplify this part of the script - you could just provide the env var with the tool name in the task list and not have to have tool specific sections. You could have a map at the top of the file that maps tools to their required volume mount strings

Reply:
Can you elaborate further on this?

Reply:
So imagine I update the GTDB image so `$CONDA_ENV` is expected to be `gtdbtk-3.0.0` or whatever. This script will send `gtdbtk-2.1.1` to the docker container which will effectively overwrite `$CONDA_ENV` and... well I'm not sure what'll happen but probably nothing good and it'll be confusing for the user.

What I'm saying is that this script shouldn't provide a command line to overwrite `ENTRYPOINT` in the docker file and handle it like the service does - call an entrypoint script have that script use environment variables to figure out what to do. In this case you could probably just put the tool name(s) in an environment variable in the task script and have the entrypoint script use that env var as an argument to the compute script.

If that's still not clear let's zoom quickly

Reply:
That also means you don't have to have an if block for each tool - just maintain a mapping of the tool name to the volume mounts as a module variable and generate the task list using that

Reply:
E.g. https://github.com/kbase/collections/blob/main/Dockerfile#L20

Reply:
👍

Reply:
:+1: